libtenzir/builtins/operators/unroll.cpp

```diff
@@ -149,40 +154,86 @@ class unroll_operator final : public crtp_operator<unroll_operator> {
 public:
   unroll_operator() = default;
 
-  explicit unroll_operator(std::string field) : field_{std::move(field)} {
+  explicit unroll_operator(ast::simple_selector field)
+    : field_{std::move(field)} {
+  }
+
+  explicit unroll_operator(located<std::string> field)
+    : field_{std::move(field)} {
   }
 
   auto
   operator()(generator<table_slice> input, operator_control_plane& ctrl) const
     -> generator<table_slice> {
-    TENZIR_UNUSED(ctrl);
+    const auto get_offset = field_.match<
+      std::function<auto(const table_slice&)->std::optional<offset>>>(
+      [&](const located<std::string>& field) {
+        return [&](const table_slice& slice) -> std::optional<offset> {
+          auto offsets = collect(slice.schema().resolve(field.inner));
+          if (offsets.empty()) {
+            diagnostic::warning("field `{}` not found", field.inner)
+              .primary(field)
+              .emit(ctrl.diagnostics());
+            return {};
+          }
+          if (offsets.size() > 1) {
+            diagnostic::warning("field `{}` resolved multiple times for `{}` "
+                                "and will be ignored",
+                                field.inner, slice.schema().name())
+              .primary(field)
+              .emit(ctrl.diagnostics());
+            return {};
+          }
+          const auto& field_type
+            = caf::get<record_type>(slice.schema()).field(offsets.front()).type;
+          if (not caf::holds_alternative<list_type>(field_type)) {
+            diagnostic::warning("expected `list`, but got `{}`",
+                                field_type.kind())
+              .primary(field)
+              .emit(ctrl.diagnostics());
+            return {};
+          }
+          return offsets.front();
+        };
+      },
+      [&](const ast::simple_selector& field) {
+        return [&](const table_slice& slice) {
+          return resolve(field, slice.schema())
+            .match(
```

Comment:
@jachris I'm facing a compilation error here that I cannot wrap my mind around right now. Likely I'm just not seeing it right now, but I could use a second pair of eyes.

```
❯ cmake --build build --target libtenzir/CMakeFiles/libtenzir_builtins.dir/builtins/operators/unroll.cpp.o
[0/2] Re-checking globbed directories...
[5/6] Building CXX object libtenzir/CMakeFiles/libtenzir_builtins.dir/builtins/operators/unroll.cpp.o
FAILED: libtenzir/CMakeFiles/libtenzir_builtins.dir/builtins/operators/unroll.cpp.o
ccache /opt/homebrew/opt/llvm@17/bin/clang++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_URL_DYN_LINK -DBOOST_URL_NO_LIB -DCAF_ENABLE_TYPE_ID_CHECKS -DFMT_SHARED -DSIMDJSON_THREADS_ENABLED=1 -DSPDLOG_COMPILED_LIB -DSPDLOG_FMT_EXTERNAL -DSPDLOG_SHARED_LIB -DTENZIR_ENABLE_BUILTINS -DXXH_ACCEPT_NULL_INPUT_POINTER=1 -DXXH_INLINE_ALL=1 -I/Users/dominiklohmann/Desktop/tenzir/unroll/build/libtenzir/include -I/Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/include -I/Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/aux/robin-map/include -isystem /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/aux/caf/libcaf_core -isystem /Users/dominiklohmann/Desktop/tenzir/unroll/build/libtenzir/aux/caf/libcaf_core -isystem /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/aux/caf/libcaf_io -isystem /Users/dominiklohmann/Desktop/tenzir/unroll/build/libtenzir/aux/caf/libcaf_io -isystem /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/aux/caf/libcaf_openssl -isystem /Users/dominiklohmann/Desktop/tenzir/unroll/build/libtenzir/aux/caf/libcaf_openssl -isystem /opt/homebrew/opt/openssl/include -isystem /Users/dominiklohmann/Desktop/tenzir/unroll/build/libtenzir/libtenzir-fbs/include -isystem /opt/homebrew/include -isystem /opt/homebrew/opt/llvm@17/include/c++/v1 -O3 -DNDEBUG -std=gnu++20 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX15.1.sdk -mmacosx-version-min=15.1 -ftemplate-backtrace-limit=0 -fno-omit-frame-pointer -Wno-unknown-warning-option -Wall -Wextra -pedantic -Werror=switch -Werror=odr -Wundef -fcolor-diagnostics -fdebug-prefix-map=/Users/dominiklohmann/Desktop/tenzir/unroll=. -MD -MT libtenzir/CMakeFiles/libtenzir_builtins.dir/builtins/operators/unroll.cpp.o -MF libtenzir/CMakeFiles/libtenzir_builtins.dir/builtins/operators/unroll.cpp.o.d -o libtenzir/CMakeFiles/libtenzir_builtins.dir/builtins/operators/unroll.cpp.o -c /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp
In file included from /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:9:
In file included from /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/include/tenzir/argument_parser.hpp:11:
In file included from /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/include/tenzir/concept/convertible/data.hpp:11:
In file included from /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/include/tenzir/fwd.hpp:11:
In file included from /Users/dominiklohmann/Desktop/tenzir/unroll/build/libtenzir/include/tenzir/config.hpp:4:
In file included from /opt/homebrew/opt/llvm@17/include/c++/v1/optional:1682:
In file included from /opt/homebrew/opt/llvm@17/include/c++/v1/iterator:683:
In file included from /opt/homebrew/opt/llvm@17/include/c++/v1/__iterator/common_iterator.h:31:
/opt/homebrew/opt/llvm@17/include/c++/v1/variant:532:20: error: constexpr variable '__fmatrix' must be initialized by a constant expression
  532 |     constexpr auto __fmatrix =
      |                    ^
  533 |         __make_fmatrix<_Visitor&&,
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~
  534 |                        decltype(_VSTD::forward<_Vs>(__vs).__as_base())...>();
      |                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/homebrew/opt/llvm@17/include/c++/v1/variant:642:20: note: in instantiation of function template specialization 'std::__variant_detail::__visitation::__base::__visit_alt<std::__variant_detail::__visitation::__variant::__value_visitor<tenzir::detail::overload<(lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:203:15), (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:215:15)>>, std::__variant_detail::__impl<tenzir::offset, tenzir::resolve_error>>' requested here
  642 |     return __base::__visit_alt(
      |                    ^
/opt/homebrew/opt/llvm@17/include/c++/v1/variant:661:12: note: in instantiation of function template specialization 'std::__variant_detail::__visitation::__variant::__visit_alt<std::__variant_detail::__visitation::__variant::__value_visitor<tenzir::detail::overload<(lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:203:15), (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:215:15)>>, tenzir::variant<tenzir::offset, tenzir::resolve_error>>' requested here
  661 |     return __visit_alt(
      |            ^
/opt/homebrew/opt/llvm@17/include/c++/v1/variant:1759:21: note: in instantiation of function template specialization 'std::__variant_detail::__visitation::__variant::__visit_value<tenzir::detail::overload<(lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:203:15), (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:215:15)>, tenzir::variant<tenzir::offset, tenzir::resolve_error>>' requested here
 1759 |   return __variant::__visit_value(_VSTD::forward<_Visitor>(__visitor),
      |                     ^
/Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/include/tenzir/variant.hpp:45:17: note: in instantiation of function template specialization 'std::visit<tenzir::detail::overload<(lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:203:15), (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:215:15)>, tenzir::variant<tenzir::offset, tenzir::resolve_error>, void>' requested here
   45 |     return std::visit(detail::overload{std::forward<Fs>(fs)...},
      |                 ^
/Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/include/tenzir/variant.hpp:178:20: note: in instantiation of function template specialization 'tenzir::detail::match<void, tenzir::variant<tenzir::offset, tenzir::resolve_error>, (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:203:15), (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:215:15)>' requested here
  178 |     return detail::match<Result>(std::move(*this), std::forward<Fs>(fs)...);
      |                    ^
/Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:202:14: note: in instantiation of function template specialization 'tenzir::variant<tenzir::offset, tenzir::resolve_error>::match<void, (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:203:15), (lambda at /Users/dominiklohmann/Desktop/tenzir/unroll/libtenzir/builtins/operators/unroll.cpp:215:15)>' requested here
  202 |             .match(
      |              ^
/opt/homebrew/opt/llvm@17/include/c++/v1/variant:532:20: note: pointer to a consteval declaration is not a constant expression
  532 |     constexpr auto __fmatrix =
      |                    ^
/opt/homebrew/opt/llvm@17/include/c++/v1/variant:571:37: note: declared here
  571 |     static constexpr decltype(auto) __dispatch(_Fp __f, _Vs... __vs) {
      |                                     ^
1 error generated.
ninja: build stopped: subcommand failed.
```

Reply:
I've also have that error every now and then... It's incredibly bad. I'm not sure why this is not getting addressed, this is a serious usability issue with `std::visit`. The best solution I have found is to do binary search by deleting code until it compiles... ☹️ I believe the error is way better on GCC.

Reply:
Figured it out, thanks for the help!